% Copyright (c) 2017 Peter A. Audano III
% GNU Free Documentation License Version 1.3 or later
% See the file COPYING.DOC for copying conditions.

\section{Introduction}
\label{sec.intro}

\subsection{About Kestrel}
\label{sec.intro.aboutkestrel}

Kestrel is an alignment-free reference-guided variant caller. Given a reference sequence and sequence reads from an isolate, Kestrel identifies variations between the reference and the isolate. Most reference sequences are contained in one or more FASTA files, and they are long assembled representations of an organisms genome. Sequence reads if an individual, or isolate, are typically contained in one or more FASTQ files.

Kestrel can identify both single nucleotide polymorphism (SNP) and insertion/deletion (indel) variants without relying on sequence reads. Most variant-calling software reads alignments generated by other tools, such as BWA or bowtie. Kestrel reads directly from the FASTQ files and skips the genome alignment step. Although it uses some algorithms based on alignments, Kestrel is alignment-free because it never attempts to align the sequence reads to the reference.

The approach Kestrel takes is to first convert the sequence reads to k-mers, which are short overlapping fragments of a given length ($k$). For example, the 4-mers in ``AACCGG'' are ``AACC'', ``ACCG'', and ``CCGG''. By default, Kestrel uses a k-mer size of 31. Because it can communicate directly with KAnalyze, Kestrel can transform the sequence reads very quickly and begin the variant-calling process.

Kestrel has several advantages over traditional alignment-guided variant callers. In some cases, it may be faster because good alignments can take a long time build. Most importantly, it can process variant-dense regions were alignments fail, and it can identify arbitrarily large insertions.

When alignment techniques work correctly, they can produce more statistically-significant results. For Kestrel to function without relying on alignments, some of the original context is lost. For example, alignments can be improved with paired-end reads, but the paired-end context is lost in k-mer space. We recommend using Kestrel as a fast first-pass variant caller or in circumstances where alignments fail or where large insertions are suspected. In some pipelines, both alignment-guided and alignment-free approaches may be used together to find all possible variants.
