% Copyright (c) 2017 Peter A. Audano III
% GNU Free Documentation License Version 1.3 or later
% See the file COPYING.DOC for copying conditions.

\section{Introduction}
\label{sec.intro}

\subsection{About This Document}
\label{sec.intro.aboutdoc}

This document is a work-in-progress. In its current form, it is incomplete and unpolished. We apologize for any inconvenience this causes, and we are actively working to get it up to our standards.


\subsection{About Kestrel}
\label{sec.intro.aboutkestrel}

Kestrel is a first-in-class variant calling toolkit that uses k-mer frequencies to detect regions of variation and resolve variants. Standard approaches require mapping sequence reads to a reference genome and searching for reads that differ from the reference. When alignments fail, sequence reads are often assembled or variants are called directly from a De~bruijn graph, which is computationally expensive. With similar resource requirements as the alignment approach, Kestrel can resolve variation in regions where quality alignments cannot be made.

Sequence reads are first decomposed into k-mers, which are short overlapping fragments of the same length. With features new in KAnalyze~\citep{Audano2014} version 2.0.0, Kestrel keeps an on-disk database of frequencies that can be rapidly searched with minimal memory requirements. The reference sequence is also decomposed into a set of k-mers and left in order. With novel algorithms to simplify searching k-mer space, the k-mer database is used to detect and resolve regions of variation.

Kestrel can identify both single nucleotide polymorphism (SNP) and insertion/deletion (indel) variants using evidence in k-mer frequencies. To our knowledge, no k-mer approach has ever been shown to efficiently recover variation over dense SNPs or large indels. These regions of variation may be larger than k-mers and the sequence reads they came from. The Kestrel algorithm seeds on the flanks of these regions and extends into them in order to resolve variation.

For full details on the Kestrel algorithm and implementation, see the publication in Oxford Bioinformatics.

Audano, P. A., Ravishankar, S., \& Vannberg, F. O. (2018). Mapping-free variant calling using haplotype\\
reconstruction from k-mer frequencies. Bioinformatics, 34(10), 1659â€“1665.\\
https://doi.org/10.1093/bioinformatics/btx753
